### PR TITLE
Improve audit view for Investment Projects

### PR DESCRIPTION
**Pull request by rachaelcodes**
## Description of change
Uses the existing Edit-History component to display the investment project audit data returned by the api. This is needed because the current audit view for investment projects only showed limited data. 

Also moves the EditHistory component and required data transformers to generic directories.

## Test instructions
* Visit Investments, click on a project, then select 'Audit Summary' from the left-hand bar
* You should see an updated audit history (as in the screenshot below).
* If there are no changes, make changes to the investment project then check the audit history again.

### Before
Screenshot_2020-04-21_at_11 50 42

### After
Screenshot 2020-04-27 at 09 53 10

## Checklist

[//]: # "When submitting a PR make sure the code review guidelines have been satisfied.
https://github.com/uktrade/data-hub-frontend/blob/master/docs/Code%20review%20guidelines.md"

- [x] Has the branch been rebased to master?
- [x] Automated tests (Any of the following when applicable: Unit, Functional or Acceptance)
- [x] Manual compatibility testing (Browsers: Chrome, Firefox, IE11, Safari)

### CODE REVIEW

File: src/apps/investments/router-projects.js [hunk 1]
```diff
@@ -131,7 +131,8 @@ router.get(
   archive.unarchiveInvestmentProjectHandler
 )
 
-router.get('/:investmentId/audit', audit.getInvestmentAudit)
+router.get('/:investmentId/audit', audit.renderProjectsView)
+router.get('/:investmentId/audit/data', audit.fetchProjectsHistoryHandler)
```
[the remark this answers is not in the document]
Comment (paulgain) [reply src/apps/investments/router-projects.js:135]: For consistency with companies we should name the path `edit-history`. The vertical navigation should also say "Edit History" - I've checked this with Rogier.
`
router.get('/:investmentId/edit-history', audit.renderProjectsView)
router.get('/:investmentId/edit-history/data', audit.fetchProjectsHistoryHandler)
`
https://github.com/uktrade/data-hub-frontend/blob/master/src/lib/urls.js#L118-L121

File: src/apps/transformers.js [hunk 2]
```diff
@@ -91,6 +106,48 @@ function transformDateStringToDateObject(dateString) {
...
+      newValue: unwrapFromArray(changes[fieldName][1]),
+    }))
+}
+
+const transformAuditLogItem = ({ timestamp, user, changes }) => {
+  const changedBy = user
+    ? isEmpty(user.name)
+      ? user.email
+      : user.name
+    : AUTOMATIC_UPDATE
```
[the remark this answers is not in the document]
Comment (paulgain) [reply src/apps/transformers.js:138]: 👍

File: src/apps/investments/router-projects.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (web-bert) [reply src/apps/investments/router-projects.js:135]: But if course using the `urls` module ;-)

File: src/apps/investments/controllers/audit.js [hunk 3]
```diff
@@ -1,37 +1,40 @@
...
-const { formatLongDate } = require('../../../common/date')
 const { getInvestmentProjectAuditLog } = require('../repos')
+const { transformAuditLog } = require('../../transformers')
 
-function formatAuditLog(logEntry) {
-  return {
-    name: get(logEntry, 'user.name'),
-    timestamp: formatLongDate(logEntry.timestamp),
-    changes: (logEntry.changes && Object.keys(logEntry.changes).length) || 0,
+const renderProjectsView = async (req, res, next) => {
```
[the remark this answers is not in the document]
Comment (paulgain) [reply src/apps/investments/controllers/audit.js:6]: Do you mine also removing `async` from here:
https://github.com/uktrade/data-hub-frontend/blob/feature/improve-investment-project-audit/src/apps/companies/apps/edit-history/controller.js#L5

File: src/apps/transformers.js [hunk 4]
```diff
@@ -91,6 +106,48 @@ function transformDateStringToDateObject(dateString) {
...
+
+  return {
+    timestamp,
+    changedBy,
+    changes: transformChanges(changes),
+  }
+}
+
+const transformAuditLog = (auditLog) => {
+  return auditLog.map((item) => transformAuditLogItem(item))
```
[the remark this answers is not in the document]
Comment (paulgain) [reply src/apps/transformers.js:148]: 👍

File: src/lib/urls.js [hunk 5]
```diff
@@ -221,6 +221,10 @@ module.exports = {
       index: url('/investments', '/profiles'),
       data: url('/investments', '/profiles/data'),
     },
+    getHistory: {
```
[the remark this answers is not in the document]
Comment (paulgain) [reply src/lib/urls.js:224]: @web-bert we want consistency with companies so this should be called `editHistory`

[the remark this answers is not in the document]
Comment (web-bert) [reply src/lib/urls.js:224]: Sounds good to me 👍

File: src/apps/investments/router-projects.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (paulgain) [reply src/apps/investments/router-projects.js:135]: Absolutely Graham :smiley:

File: src/apps/constants.js [hunk 6]
```diff
@@ -11,8 +11,43 @@ const EXPORT_INTEREST_STATUS = {
...
+  pending_dnb_investigation: 'Pending D&B investigation',
+  registered_address_1: 'Registered address line 1',
+  registered_address_2: 'Registered address line 2',
+  sector: 'DIT sector',
+  trading_names: 'Trading name(s)',
+  uk_region: 'DIT region',
+  uk_based: 'UK based',
+  vat_number: 'VAT number (optional)',
+  website: "Company's website (optional)",
+}
```
[the remark this answers is not in the document]
Comment (paulgain) [reply src/apps/constants.js:44]: These constants are specific for companies and companies only, you cannot edit these fields within Investments. You should create your own specific constants for Investments and move this back to its original location. It makes no sense/reusing this.

File: src/apps/transformers.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (paulgain) [reply src/apps/transformers.js:138]: Automatic updates refer to Dunn & Bradstreet company updates **only**, you will not get automatic updates for editable fields with an Investment company. This should be put pack to its original location.

File: src/apps/__test__/transformers.test.js [hunk 7]
```diff
@@ -187,3 +189,265 @@ describe('Global transformers', () => {
...
+          {
+            fieldName: 'Address line 1',
+            oldValue: '1600 Amphitheatre Pkwyz',
+            newValue: '1600 Amphitheatre Pkwy',
+          },
+        ],
+      },
+    ])
+  })
+})
```
Comment (paulgain) [inline src/apps/__test__/transformers.test.js:453]: All of these tests are very specific to editing a company, they should to be moved back to their original location.

**Comment by paulgain** [comment]
@rachaelcodes I highly recommend you edit every single field with an Investment Project. Then you will see how your view looks. You will clearly see the changes that need to be made and how that relates to the code. You will also be able to test the pagination. It's important to note that any code that is specific to companies should remain in companies, it makes no sense in pulling that out. We only want generic/reusable code within `src/client/components/EditHistory`. The same applies to both unit and functional tests.

File: src/apps/companies/apps/edit-history/client/EditCompanyHistory.jsx [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+import {
+  ARCHIVED,
+  NOT_ARCHIVED,
+  YES,
+  NO,
+  AUTOMATIC_UPDATE,
+  CHANGE_TYPE_TEXT,
+} from '../constants'
+import { NOT_SET } from '../../../../../client/components/EditHistory/constants'
```
Comment (paulgain) [inline src/apps/companies/apps/edit-history/client/EditCompanyHistory.jsx:18]: We should delete the `.../client/components/EditHistory/constant.js` file considering there are only two constants.

```javascript
const NOT_SET = 'Not set'
const DEFAULT_ITEMS_PER_PAGE = 10
```
Both should be moved to the constant files of Investments and Companies, it has the benefit of only importing a single constant file into both files. The duplication is minimal and worth the trade off.

File: src/apps/investments/transformers/edit-history.js [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+  return {
+    timestamp,
+    changedBy,
+    changes: transformChanges(changes),
+  }
+}
+
+const transformAuditLog = (auditLog) => {
+  return auditLog.map((item) => transformAuditLogItem(item))
```
Comment (paulgain) [inline src/apps/investments/transformers/edit-history.js:37]: This should be: `return auditLog.map(transformAuditLogItem)`

**Pull request merged**